Add Telemetry guide (#644)
akka/akka-platform-guide: docs-source/docs/modules/telemetry/pages/getting-started.adoc

= Getting started
:page-toclevels: 3

include::ROOT:partial$include.adoc[]

== Lightbend Account & Credentials

ifdef::todo[TODO: Support has a signup process that will get AWS customers Lightbend credentials to download commercial dependencies.]

== Set up Cinnamon Agent

To get started, you should configure {telemetry}/setup/setup.html[your build] to use the Cinnamon Agent for one of the following build systems: {telemetry}/setup/cinnamon-agent-sbt.html[sbt], {telemetry}/setup/cinnamon-agent-maven.html[Maven], or {telemetry}/setup/cinnamon-agent-gradle.html[Gradle].

=== Maven build changes

To automatically pull down the agent as part of your build, you add the following plugin to download as part of the `package` task.

ifdef::todo[TODO: Externalize pom.xml configuration to a sample application?]

```xml

> <plugin>

Reviewer: We can include this in the shopping example projects, but commented out. Then you can pull in this as snippets. Would also be good to include `sbt` since we have that in other places for Scala.

            <overWrite>true</overWrite>
            <destFileName>cinnamon-agent.jar</destFileName>
          </artifactItem>
        </artifactItems>
      </configuration>
    </execution>
  </executions>
</plugin>
```

Next, the agent needs to be added as a java command-line option, for example, here is how to add it for the `exec:exec` task:

```xml
<plugin>
  <groupId>org.codehaus.mojo</groupId>
  <artifactId>exec-maven-plugin</artifactId>
  <version>1.6.0</version>
  <configuration>
    <executable>java</executable>
    <arguments>
      <argument>-classpath</argument>
      <classpath/>
      <argument>-javaagent:${project.build.directory}/dependency/cinnamon-agent.jar</argument>
      <argument>... your other arguments here ...</argument>
    </arguments>
  </configuration>
</plugin>
```

See full details and a complete sample {telemetry}/setup/cinnamon-agent-maven.html[here].

== Instrumenting your application

Lightbend Telemetry can capture data for {akka-http}/introduction.html[Akka HTTP and Akka gRPC] related features. After adding the Cinnamon Agent as described above, make sure that you add the Cinnamon Akka HTTP module dependency to your build file:

> ```xml

Reviewer: same with dependencies, good to include in examples (as commented out sections)

<!-- Use Akka HTTP instrumentation -->
<dependency>
  <groupId>com.lightbend.cinnamon</groupId>
  <artifactId>cinnamon-akka-http_2.12</artifactId>
  <version>2.16.0-SNAPSHOT</version>
</dependency>
<!-- Use Akka gRPC instrumentation -->
<dependency>
  <groupId>com.lightbend.cinnamon</groupId>
  <artifactId>cinnamon-akka-grpc_2.12</artifactId>
  <version>2.16.0-SNAPSHOT</version>
</dependency>
```

== Telemetry Configuration

A {telemetry}/instrumentations/akka-http/akka-http-configuration.html#example-configurations[simple configuration] for your Akka HTTP server would be the one that matches all paths, hosts and ports:

ifdef::todo[TODO: Externalize configuration snippets to a sample application?]

```conf

> cinnamon.akka.http.servers {

Reviewer: conf can be included in samples

  "*:*" {
    paths {
      "*" {
        metrics = on
      }
    }
  }
}
```

A similar configuration can be done for the client:

```conf
cinnamon.akka.http.clients {
  "*:*" {
    paths {
      "*" {
        metrics = on
      }
    }
  }
}
```

== Whats Next?

> The next step is to use Prometheus as a metrics backend.

Reviewer: "next" links are included automatically

= Getting started
:page-toclevels: 3

> include::ROOT:partial$include.adoc[]

Reviewer 2: I will be adding that section info

> include::ROOT:partial$include.adoc[]

Reviewer 2: added

== Lightbend Account & Credentials

> Use of lightbend Telemetry requires an account and credentials. Account options are described on the opening page of this xref::ROOT:index.adoc[].

Reviewer 3: ```suggestion
Use of lightbend Telemetry requires an account and credentials. Account options are described on the opening page of this xref:ROOT:index.adoc[].
```

The syntax `xref::module:page.adoc[]` is not valid for all AsciiDoc interpreters. On My machine, the `xref::ROOT:index.adoc[]` syntax is rendered on my IDE but not by the antora-to-HTML renderer.

Reviewer 3: The sentence `Account options are described on the opening page of this xref::ROOT:index.adoc[].` links to a page that has no information about Accounts. Is the link pointing to the correct page?

Reviewer 2: Yes, it is the correct page. The 2nd, 3rd, and 4rth, paras talk and link to the subscription options. I revise the text to indicate subscriptions.

== Set up Cinnamon Agent

> 

Reviewer 2: I'll add in reference to using an IDE

> To get started, you should configure {cinnamon-docs}/setup/setup.html[your build {tab-icon}, window="tab"] to use the Cinnamon Agent for one of the following build systems: {cinnamon-docs}/setup/cinnamon-agent-sbt.html[SBT {tab-icon}, window="tab"], {cinnamon-docs}/setup/cinnamon-agent-maven.html[Maven {tab-icon}, window="tab"], or {cinnamon-docs}/setup/cinnamon-agent-gradle.html[Gradle {tab-icon}, window="tab"].

Reviewer 3: ```suggestion
To get started, you should configure {cinnamon-docs}/setup/setup.html[your build {tab-icon}, window="tab"] to use the Cinnamon Agent for one of the following build systems: {cinnamon-docs}/setup/cinnamon-agent-sbt.html[sbt {tab-icon}, window="tab"], {cinnamon-docs}/setup/cinnamon-agent-maven.html[Maven {tab-icon}, window="tab"], or {cinnamon-docs}/setup/cinnamon-agent-gradle.html[Gradle {tab-icon}, window="tab"].
```

I think we always use `sbt` (all lowercase).

> To get started, you should configure {cinnamon-docs}/setup/setup.html[your build {tab-icon}, window="tab"] to use the Cinnamon Agent for one of the following build systems: {cinnamon-docs}/setup/cinnamon-agent-sbt.html[SBT {tab-icon}, window="tab"], {cinnamon-docs}/setup/cinnamon-agent-maven.html[Maven {tab-icon}, window="tab"], or {cinnamon-docs}/setup/cinnamon-agent-gradle.html[Gradle {tab-icon}, window="tab"].

Reviewer 2: Per our discussion, I think will stay with some enhancement

> === Maven build changes

Reviewer 3: This section demonstrates the required changes for a Maven build. It should include, at least, `sbt` aswell.

To automatically pull down the agent as part of your build, you add the following plugin to download as part of the `package` task.

ifdef::todo[TODO: Externalize pom.xml configuration to a sample application?]

> [source]
> ----
> xml

Reviewer 3: ```suggestion
.pom.xml:
[source,xml,indent=0]
----
```

<plugin>
  <groupId>org.apache.maven.plugins</groupId>
  <artifactId>maven-dependency-plugin</artifactId>
  <version>3.1.1</version>
  <executions>
    <execution>
      <id>copy</id>
      <phase>compile</phase>
      <goals>
        <goal>copy</goal>
      </goals>
      <configuration>
        <artifactItems>
          <artifactItem>
            <groupId>com.lightbend.cinnamon</groupId>
            <artifactId>cinnamon-agent</artifactId>
            <version>2.15.0</version>
            <overWrite>true</overWrite>
            <destFileName>cinnamon-agent.jar</destFileName>
          </artifactItem>
        </artifactItems>
      </configuration>
    </execution>
  </executions>
</plugin>
----

Next, the agent needs to be added as a java command-line option. The following is an example of how to add the agent for the `exec:exec` task:

> [source,xml]

Reviewer 3: ```suggestion
.pom.xml:
[source,xml]
```

----
<plugin>
  <groupId>org.codehaus.mojo</groupId>
  <artifactId>exec-maven-plugin</artifactId>
  <version>1.6.0</version>
  <configuration>
    <executable>java</executable>
    <arguments>
      <argument>-classpath</argument>
      <classpath/>
      <argument>-javaagent:${project.build.directory}/dependency/cinnamon-agent.jar</argument>
      <argument>... your other arguments here ...</argument>
    </arguments>
  </configuration>
</plugin>
----

> The {cinnamon-docs}/setup/cinnamon-agent-maven.html[Full Telemetry Documentation {tab-icon}, window="tab"] contains full details and a complete sample.

Reviewer 3: ```suggestion
The {cinnamon-docs}/setup/cinnamon-agent-maven.html[Telemetry Documentation {tab-icon}, window="tab"] contains full details and a complete sample.
```

`full ` used twice.

Reviewer 2: Actually, I think we should keep the first 'full' and change the second 'full' to 'more.'

== Instrumenting your application

> Lightbend Telemetry can capture data for {akka-http}/introduction.html[Akka HTTP and Akka gRPC {tab-icon}, window="tab"] related features. After adding the Cinnamon Agent as described above, make sure that you add the Cinnamon Akka HTTP module dependency to your build file:

Reviewer 3: ```suggestion
Imagine we're instrumenting a simple application where we only wanted to capture metrics and traces for {akka-http}/introduction.html[Akka HTTP and Akka gRPC {tab-icon}, window="tab"]. After adding the Cinnamon Agent as described above, make sure that you add the Cinnamon Akka HTTP module dependency to your build file:
```

Reviewer 3: The original write-up has the same "HTTP and gRPC" vibes I mentioned above. The suggestion I made yesterday no longer makes sense with all the added improvements.

[source,xml]
----
<!-- Use Akka HTTP instrumentation -->
<dependency>
  <groupId>com.lightbend.cinnamon</groupId>
  <artifactId>cinnamon-akka-http_2.12</artifactId>
  <version>2.16.0-SNAPSHOT</version>
</dependency>
<!-- Use Akka gRPC instrumentation -->
<dependency>
  <groupId>com.lightbend.cinnamon</groupId>
  <artifactId>cinnamon-akka-grpc_2.12</artifactId>
  <version>2.16.0-SNAPSHOT</version>
</dependency>
----

== Telemetry Configuration

A {cinnamon-docs}/instrumentations/akka-http/akka-http-configuration.html#example-configurations[simple configuration {tab-icon}, window="tab"] for your Akka HTTP server will match all paths, hosts and ports:

ifdef::todo[TODO: Externalize configuration snippets to a sample application?]

> [source,conf]

Reviewer 3: ```suggestion
.src/main/resources/telemetry.conf:
[source,hocon]
```

----
cinnamon.akka.http.servers {
  "*:*" {
    paths {
      "*" {
        metrics = on
      }
    }
  }
}
----

A similar configuration can be done for the client:

> [source,conf]

Reviewer 3: ```suggestion
.src/main/resources/telemetry.conf:
[source,hocon]
```

= Getting started
:page-toclevels: 3
:page-supergroup-java-scala: Language

include::ROOT:partial$include.adoc[]

> Lightbend Telemetry can capture data for {akka-http}/introduction.html[Akka HTTP and Akka gRPC {tab-icon}, window="tab"] related features.

Reviewer 3: Same as above.

Lightbend provides two essential pieces for Monitoring your deployments; Telemetry (a.k.a. Cinnamon) and Lightbend Console. This documentation is focused solely on Telemetry. Telemetry makes it possible to gather metric, event and trace information from your applications. The information is transferred to various backends. The rest of this documentation provides an overview of Telemetry and explains how to get started, configure and run Telemetry in production.

:sectnums:

== Subscriptions and credentials

> Use of lightbend Telemetry requires a subscription and credentials. A Lightbend subscription includes open source components that are distributed from the https://maven.apache.org/repository/[Maven Central Repository], as well as the commercial components which are distributed from a private repository. A https://www.lightbend.com/lightbend-platform-subscription[Lightbend Subscription] entitles you to use all commercial features and includes https://www.lightbend.com/lightbend-platform-subscription[support] and https://academy.lightbend.com/dashboard/[Lightbend Academy].

Reviewer 3: ```suggestion
Use of Lightbend Telemetry requires a subscription and credentials. A Lightbend subscription includes open source components that are distributed from the https://maven.apache.org/repository/[Maven Central Repository], as well as the commercial components which are distributed from a private repository. A https://www.lightbend.com/lightbend-platform-subscription[Lightbend Subscription] entitles you to use all commercial features and includes https://www.lightbend.com/lightbend-platform-subscription[support] and access to the https://academy.lightbend.com/dashboard/[Lightbend Academy].
```

> Telemetry is also available by subscription to the Akka Cloud Platform which is via {aws-marketplace}[Amazon Web Services (AWS) {tab-icon}, window="tab"] or {gcp-marketplace}[Google Cloud Platform (GCP) {tab-icon}].

Reviewer 3: ```suggestion
Lightbend Telemetry is also available by subscription to the Akka Cloud Platform which is via {aws-marketplace}[Amazon Web Services (AWS) {tab-icon}, window="tab"] or {gcp-marketplace}[Google Cloud Platform (GCP) {tab-icon}].
```

Once you have a subscription, visit https://www.lightbend.com/account/lightbend-platform/credentials[lightbend.com] and login to learn how to add credentials to your development projects.

== Set up Cinnamon Agent

To get started, you should configure {cinnamon-docs}/setup/setup.html[your build {tab-icon}, window="tab"] to use the Cinnamon Agent for one of the following build systems: {cinnamon-docs}/setup/cinnamon-agent-sbt.html[sbt {tab-icon}, window="tab"], {cinnamon-docs}/setup/cinnamon-agent-maven.html[Maven {tab-icon}, window="tab"], or {cinnamon-docs}/setup/cinnamon-agent-gradle.html[Gradle {tab-icon}, window="tab"].

=== Commercial Repositories

Here we assume that you have the repositories configured as environment variables that can be read by your build tool:

[source,shell script]
----
export LIGHTBEND_COMMERCIAL_MVN=https://repo.lightbend.com/...
export LIGHTBEND_COMMERCIAL_IVY=https://repo.lightbend.com/...

> ----

Reviewer 3: Nice! It's great to present this as a good practice!

----
+
And then configure it in your project:
+
.project/plugins.sbt
[source,scala,indent=0]
----
include::example$telemetry-sample/build.sbt[tag=telemetry-javaagent-docker]
----

== Telemetry Configuration

Lightbend Telemetry supports a {cinnamon-docs}/plugins/chmetrics/reporters.html#reporters[reporter {tab-icon} window="tab"] that can be used to see the metrics locally in your terminal. 

.src/main/resources/telemetry.conf
[source,hocon]
----
include::example$telemetry-sample/src/main/resources/telemetry.conf[tag=telemetry-config-console]
----

A {cinnamon-docs}/instrumentations/akka-http/akka-http-configuration.html#example-configurations[simple configuration {tab-icon}, window="tab"] for your Akka HTTP server will match all paths, hosts and ports:

.src/main/resources/telemetry.conf
[source,hocon]
----
include::example$telemetry-sample/src/main/resources/telemetry.conf[tag=telemetry-config-http-server]
----

A similar configuration can be done for the client:

.src/main/resources/telemetry.conf
[source,hocon]
----
include::example$telemetry-sample/src/main/resources/telemetry.conf[tag=telemetry-config-http-client]
----

> 

Reviewer 3: Maybe we can add a final section mentioning deployment and add a cross-reference to the Akka platform operator (https://developer.lightbend.com/docs/akka-platform-guide/deployment/operator-reference.html#_reporting_metrics_to_prometheus)
